### PR TITLE
Addressing shapeshifting issues with window functions

### PR DESCRIPTION
**Pull request by somu-imply**
Some window function queries can throw this error 

Screen Shot 2024-01-31 at 10 58 21 AM
when used  in conjecture with `druid.server.http.maxSubqueryBytes=auto` or any other value e.g. "100000". For example the query context of the failed query here was
```
{
  "enableWindowing": true,
  "debug": true,
  "useCache": false,
  "populateCache": false,
  "maxSubqueryBytes": "100000"
}
```
The same issue could surface if `druid.server.http.maxSubqueryBytes=auto` (or any other value) is specified in the broker runtime properties

The reason is that in such a case the reference counting segment wraps over a frame segment and a frame segment cannot shapeshift. This PR tries to address that. 

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/rowsandcols/concrete/FrameRowsAndColumns.java [hunk 1]
```diff
@@ -91,4 +93,10 @@ public <T> T as(Class<T> clazz)
     }
     return null;
   }
+
+  @Override
+  public void close() throws IOException
+  {
+    // nothing to close
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/concrete/FrameRowsAndColumns.java:100]: I am wondering if we need to clear the cache here. @LakshSingla - maybe you know?

Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/rowsandcols/concrete/FrameRowsAndColumns.java:100]: I think since the close is called, there'd be no attempt to use the object, so we don't wanna clear the cache. 
I was wondering if there's a .close() method on the cached objects or any other resources that we are seeing, but I don't think there is.

**Comment by somu-imply** [comment]
There are some coverage failures in processing as the UT introduced is in the calcite query tests

**Comment by LakshSingla** [comment]
The error message specifies that its `RUNTIME_FAILURE (OPERATOR)`. I think that DEVELOPER is a better persona for this error if it happens again.

File: processing/src/main/java/org/apache/druid/frame/segment/FrameSegment.java [hunk 2]
```diff
@@ -77,4 +80,15 @@ public void close()
   {
     // Nothing to close.
   }
+
+  @SuppressWarnings("unchecked")
+  @Nullable
+  @Override
+  public <T> T as(@Nonnull Class<T> clazz)
+  {
+    if (CloseableShapeshifter.class.equals(clazz)) {
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/frame/segment/FrameSegment.java:89]: This code is breaking a code path where default method is called with StorageAdaptor.class

```
 default <T> T as(@Nonnull Class<T> clazz)
  {
    if (clazz.equals(QueryableIndex.class)) {
      return (T) asQueryableIndex();
    } else if (clazz.equals(StorageAdapter.class)) {
      return (T) asStorageAdapter();
    }
    return null;
  }
}
```
Is that okay ?

Comment (gianm) [reply processing/src/main/java/org/apache/druid/frame/segment/FrameSegment.java:89]: Should it be `else return super.as(clazz)` instead of `else return null`?

Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/frame/segment/FrameSegment.java:89]: +1

Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/frame/segment/FrameSegment.java:89]: I looked at the comment in 
```
* Implementations which accept classes other than {@link QueryableIndex} or {@link StorageAdapter} are limited 
   * to using those classes within the extension. This means that one extension cannot rely on the `Segment.as` 
   * behavior of another extension.
 ```
 Since here we are using a different one here, based on the comment I resorted to returning null. The same pattern is observed in https://github.com/apache/druid/blob/50bae96e8b44913faeb7c64a79430ffd7cb843a4/processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java#L78. Can you please help me understand better @gianm  @cryptoe

File: processing/src/main/java/org/apache/druid/query/operator/SegmentToRowsAndColumnsOperator.java [hunk 3]
```diff
@@ -44,7 +46,13 @@ public Closeable goOrContinue(Closeable continuation, Receiver receiver)
   {
     try (final CloseableShapeshifter shifty = segment.as(CloseableShapeshifter.class)) {
       if (shifty == null) {
-        throw new ISE("Segment[%s] cannot shapeshift", segment.getClass());
+        final Class<? extends Segment> classOfSegment;
+        if (segment instanceof ReferenceCountingSegment) {
```
Comment (kgyrtkirk) [inline processing/src/main/java/org/apache/druid/query/operator/SegmentToRowsAndColumnsOperator.java:50]: this seems like an abstraction leak - I think this should be the responsibility of `Segment` somethow.

Wouldn't it be better to implement the `toString()` methods so that it contains some details - and class name of the base segment if its a `ReferenceCountingSegment` ?

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java [hunk 4]
```diff
@@ -219,6 +219,28 @@ public void windowQueryTest() throws Exception
     }
   }
 
+  @Test
+  @SuppressWarnings("unchecked")
+  public void windowQueryTestWithCustomContextMaxSubqueryBytes() throws Exception
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java:224]: I wonder if instead of running all testcases in this mode as well - wouldn't it be better to introduce the ability to specify querycontext keys in the `sqlTest` files; so that `rank_handling.sqlTest` can set it

Comment (somu-imply) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java:224]: Good idea, I have other tests in flight as well, we can merge this in and I'll send out a refactoring for the test cases

**Comment by somu-imply** [comment]
Although the comment is there, I don't see any loss of generalization for calling the super's default method for these Segments. Updated the occurrences

**Comment by krishnat2** [comment]
Thanks to @somu-imply and @abhishekagarwal87 for a quick fix! Really appreciate it.

File: processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java [hunk 5]
```diff
@@ -81,6 +81,6 @@ public <T> T as(@Nonnull Class<T> clazz)
       return (T) new QueryableIndexRowsAndColumns(index);
     }
 
-    return null;
+    return Segment.super.as(clazz);
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java:84]: Is this change related to the PR ?

File: processing/src/main/java/org/apache/druid/segment/ArrayListSegment.java [hunk 6]
```diff
@@ -109,7 +109,7 @@ public <T> T as(Class<T> clazz)
     if (CloseableShapeshifter.class.equals(clazz)) {
       return (T) new MyCloseableShapeshifter();
     }
-    return null;
+    return Segment.super.as(clazz);
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/segment/ArrayListSegment.java:112]: Same comment here : https://github.com/apache/druid/pull/15807/files#r1478490109

File: processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java (hunk 5 above)
Comment (soumyava) [reply processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java:84]: Yes there were some missing calls to the default as, which were fixed

File: processing/src/main/java/org/apache/druid/segment/ArrayListSegment.java (hunk 6 above)
Comment (soumyava) [reply processing/src/main/java/org/apache/druid/segment/ArrayListSegment.java:112]: Same as before

File: processing/src/main/java/org/apache/druid/segment/Segment.java [hunk 7]
```diff
@@ -69,4 +69,9 @@ default <T> T as(@Nonnull Class<T> clazz)
     }
     return null;
   }
+
+  default String asString()
+  {
+    return "" + getClass();
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/segment/Segment.java:75]: nit:
```suggestion
    return getClass().toString();
```
